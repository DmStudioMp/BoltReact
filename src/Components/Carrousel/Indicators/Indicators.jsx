import React from "react";
import PropTypes from "prop-types";

export default function Indicators(props) {
  const slides = props.slides;
  const onSelect = props.onSelect;
  // Validar que slides sea un array antes de continuar
  if (!Array.isArray(slides) || slides.length === 0) {
    return null; // Si no hay slides o no es un array válido, no renderiza nada
  }

  return (
    <div className="carousel-indicators justify-content-center" role="tablist">
      {slides.map((_, index) => (
        <button
          key={index}
          type="button"
          className="indicator-btn"
          onClick={() => onSelect(index)}
          aria-label={`Go to slide ${index + 1}`}
          aria-selected={index === 0} // El primer slide puede ser seleccionado por defecto
        >
          {index + 1}
        </button>
      ))}
    </div>
  );
}

// Validación de props
Indicators.propTypes = {
  slides: PropTypes.arrayOf(PropTypes.object).isRequired,
  onSelect: PropTypes.func.isRequired, // Este prop es obligatorio
};

// Valores predeterminados
Indicators.defaultProps = {
  slides: [], // Predeterminamos un array vacío para slides
};
