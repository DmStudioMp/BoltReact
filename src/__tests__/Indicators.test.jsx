import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Indicators from "../Components/Carrousel/Indicators/Indicators.jsx";
import "@testing-library/jest-dom";

describe("Indicators Component", () => {
  const mockOnSelect = jest.fn();

  afterEach(() => {
    mockOnSelect.mockClear(); // Limpia el mock despuÃ©s de cada test
  });

  test("renders without crashing when slides are empty", () => {
    render(<Indicators slides={[]} onSelect={mockOnSelect} />);
    expect(screen.queryByRole("button")).toBeNull();
  });

  test("renders correct number of indicators", () => {
    const mockSlides = [{ id: 1 }, { id: 2 }, { id: 3 }];
    render(<Indicators slides={mockSlides} onSelect={mockOnSelect} />);

    const buttons = screen.getAllByRole("button");
    expect(buttons).toHaveLength(mockSlides.length);

    buttons.forEach((button, index) => {
      expect(button).toHaveAttribute("aria-label", `Go to slide ${index + 1}`);
      expect(button).toHaveTextContent((index + 1).toString());
    });
  });

  test("calls onSelect with correct index when a button is clicked", () => {
    const mockSlides = [{ id: 1 }, { id: 2 }, { id: 3 }];
    render(<Indicators slides={mockSlides} onSelect={mockOnSelect} />);

    const buttons = screen.getAllByRole("button");
    fireEvent.click(buttons[1]); // Click on second button

    expect(mockOnSelect).toHaveBeenCalledTimes(1);
    expect(mockOnSelect).toHaveBeenCalledWith(1);
  });

  test("does not throw an error when onSelect is not provided", () => {
    const mockSlides = [{ id: 1 }, { id: 2 }, { id: 3 }];
    render(<Indicators slides={mockSlides} onSelect={() => {}} />);
    const buttons = screen.getAllByRole("button");

    expect(() => fireEvent.click(buttons[0])).not.toThrow();
  });

  test("matches snapshot", () => {
    const mockSlides = [{ id: 1 }, { id: 2 }, { id: 3 }];
    const { container } = render(
      <Indicators slides={mockSlides} onSelect={mockOnSelect} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders default props correctly", () => {
    const { container } = render(<Indicators onSelect={mockOnSelect} />);
    expect(container.firstChild).toBeNull();
  });
});
