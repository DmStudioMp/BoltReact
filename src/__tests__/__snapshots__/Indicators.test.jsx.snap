// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicators Component matches snapshot 1`] = `
<div
  class="carousel-indicators justify-content-center"
  role="tablist"
>
  <button
    aria-label="Go to slide 1"
    aria-selected="true"
    class="indicator-btn"
    type="button"
  >
    1
  </button>
  <button
    aria-label="Go to slide 2"
    aria-selected="false"
    class="indicator-btn"
    type="button"
  >
    2
  </button>
  <button
    aria-label="Go to slide 3"
    aria-selected="false"
    class="indicator-btn"
    type="button"
  >
    3
  </button>
</div>
`;
